internal/engine/wazevo/backend/isa/arm64/lower_instr.go

```diff
@@ -395,6 +399,189 @@ func (m *machine) LowerInstr(instr *ssa.Instruction) {
 	m.FlushPendingInstructions()
 }
 
+func (m *machine) lowerVcheckTrue(instr *ssa.Instruction) {
+	x, lane := instr.ArgWithLane()
+
+	rm := m.getOperand_NR(m.compiler.ValueDefinition(x), extModeNone)
+	rd := operandNR(m.compiler.VRegOf(instr.Return()))
+	tmp := operandNR(m.compiler.AllocateVReg(regalloc.RegTypeOf(ssa.TypeV128)))
+
+	if instr.Opcode() == ssa.OpcodeVallTrue && lane == ssa.VecLaneI64x2 {
+		ins := m.allocateInstr()
+		ins.asVecMisc(vecOpCmeqZero, rd, rm, vecArrangement2D)
+		m.insert(ins)
+
+		addp := m.allocateInstr()
+		addp.asVecRRR(vecOpAddp, rd, rd, rd, vecArrangement2D)
+		m.insert(addp)
+
+		fcmp := m.allocateInstr()
+		fcmp.asFpuCmp(rd, rd, true)
+		m.insert(fcmp)
+
+		cset := m.allocateInstr()
+		cset.asCSet(rd.nr(), eq)
+		m.insert(cset)
+
+		return
+	}
+
+	ins := m.allocateInstr()
+
+	// Create a scalar value with umaxp or uminv, then compare it against zero.
+	if instr.Opcode() == ssa.OpcodeVanyTrue {
+		ins.asVecRRR(vecOpUmaxp, tmp, rm, rm, vecArrangement16B)
+	} else {
+		arr := ssaLaneToArrangement(lane)
+		ins.asVecLanes(vecOpUminv, tmp, rm, arr)
+	}
+
+	m.insert(ins)
+
+	movv := m.allocateInstr()
+	movv.asMovFromVec(rd, tmp, vecArrangementD, vecIndex(0))
+	m.insert(movv)
+
+	fc := m.allocateInstr()
+	fc.asCCmpImm(rd, uint64(0), al, 0, true)
+	m.insert(fc)
+
+	cset := m.allocateInstr()
+	cset.asCSet(rd.nr(), ne)
+	m.insert(cset)
+}
+
+func (m *machine) lowerVhighBits(instr *ssa.Instruction) {
+	x, lane := instr.ArgWithLane()
+	rm := m.getOperand_NR(m.compiler.ValueDefinition(x), extModeNone)
+	rd := operandNR(m.compiler.VRegOf(instr.Return()))
+
+	r0 := operandNR(m.compiler.AllocateVReg(regalloc.RegTypeOf(ssa.TypeI64)))
+	v0 := operandNR(m.compiler.AllocateVReg(regalloc.RegTypeOf(ssa.TypeV128)))
+	v1 := operandNR(m.compiler.AllocateVReg(regalloc.RegTypeOf(ssa.TypeV128)))
+
+	switch lane {
+
+	case ssa.VecLaneI8x16:
+		sshr := m.allocateInstr()
+		sshr.asVecShiftImm(vecOpSshr, v1, rm, operandShiftImm(7), vecArrangement16B)
+		m.insert(sshr)
+
+		// m.lowerConstantI64(r0.nr(), 0x0008000400020001)
+		m.insertMOVZ(r0.nr(), 0x0201, 0, true)
+		m.insertMOVK(r0.nr(), 0x0804, 1, true)
+		m.insertMOVK(r0.nr(), 0x2010, 2, true)
+		m.insertMOVK(r0.nr(), 0x8040, 3, true)
+
+		dup := m.allocateInstr()
+		dup.asVecDup(v0, r0, vecArrangement2D)
```

Comment:
what if v0 is used somewhere else? (seems a same bug as we discussed) Instead use the temporary value.